### PR TITLE
Eigensolver inaccuracies at low magnetic field

### PR DESCRIPTION
**Pull request by RHanley1**
When a user inputs B=0, the degenercay of the states causes the
calculation of eigenstates to crash. I have added a warning and
a specific case to change B = 0 to B = 1e-10 so that this doesn't
happen.

### CODE REVIEW

**Comment by RHanley1** [comment]
Looking into this in more detail, it is not as trivial as first thought. It seems that the eigensolver thinks it has converged, but when one calculates the quantum number 'M', you can get weird non-integer values, such as 1.7.... for example. This then leads to inaccuracies and intermittent crashing dependent on what M is rounded to using np.rint(). I thought the code below (inserted into line 334 of common.py) may deal with this. Any thoughts?

```python
self.M[lev] = M = (2*np.diag(V.conj().T@(Iz+Jz)@V))/2

            eps = np.rint(M) - M
            if True in (eps > 1e-5):
                raise ValueError('Matrix diagnolisation inaccuracies.'\
                                 ' Is B too small to lift the degeneracy'\
                                 ' of states?')
            else:
                M = np.rint(M)
```

**Comment by hartytp** [comment]
Thanks for digging into this. That makes sense: the issue here is
- the current code assumes we can find the angular momentum eignestates by finding the Hamiltonian's eigenstates
- at zero field the states within a level are completely degenerate, so any superposition of angular momentum states has the same energy. As a result, our eigensolver doesn't give us angular momentum eigenstates. So the previous assumption is not valid
- Or, to put it another way, if you have B=0 and no hyperfine structure then H is strictly zero. (What happens to the code in that case?)

If we wanted the code to still work in the case of degenerate states, I guess we could add a special case that switches to using the eigenstates of `F`, but that feels nasty (unless we ever need to handle 0 field explicitly). So, your proposed solution is to test for this case by checking that the states are all angular momentum eigenstates explicitly. That seems fine. 

Can you update the commit message and PR description to explain this more clearly?

```python
            eps = np.rint(M) - M
            if True in (eps > 1e-5):
                raise ValueError('Matrix diagnolisation inaccuracies.'\
                                 ' Is B too small to lift the degeneracy'\
                                 ' of states?')
            else:
                M = np.rint(M)
```

What about `M=1/2`?

A neater way of expressing the same logic would be:
```python
M = (2*np.diag(V.conj().T@(Iz+Jz)@V))/2
if max(abs(M-np.rint(M))) > 1e-5:
    raise ValueError('Error finding angular momentum eigenstates at {}T. Is the field too small to life the degeneracy of states?`.format(B))
self.M[lev] = np.rint(M).asint()
```

NB no `else` needed.

Otherwise, I think this makes sense as a solution.

**Comment by RHanley1** [comment]
> If we wanted the code to still work in the case of degenerate states,

I don't think we necessarily want it to work for B=0, but you can get the same issue occurring for non-zero (but small fields), and it's worth while making sure this doesn't catch people out. 

> Can you update the commit message and PR description to explain this more clearly?

Yes, just wanted to discuss before making N commits for the same thing. 

> What about `M=1/2`?
It's a good point. For the ions we care about, nuclear spin is half-integer and the electron spin is half-integer as we only have one outer-shell electron, so using rint(M) is fine to fix numerical inaccuracies 1.999... = 2. However, if others wanted to use this code for other things, such as neutral Sr, where F can be half-integer, this would break. Is it worth rounding to the nearest half-integer in the code?

**Comment by hartytp** [comment]
@rhanley1 what about 40Ca+ (I=0)? We should definitely cover the half-integer M. I bet this bug occurs in quite a few places in the code, so worth having a pr to fix that. Maybe also add a 40Ca+ test at the same time

Agreed on all other points

File: ion_phys/common.py [hunk 1]
```diff
@@ -330,7 +330,13 @@ def setB(self, B):
             self.MJax[lev] = np.kron(np.arange(-J, J + 1), np.ones(I_dim))
             self.MI[lev] = np.rint(2*np.diag(V.conj().T@(Iz)@V))/2
             self.MJ[lev] = np.rint(2*np.diag(V.conj().T@(Jz)@V))/2
-            self.M[lev] = M = np.rint(2*np.diag(V.conj().T@(Iz+Jz)@V))/2
+
+            M = (np.diag(V.conj().T@(Iz+Jz)@V))
```
Comment (hartytp) [inline ion_phys/common.py:334]: Nit pick: the outer brackets are not needed here.

**Pull request merged**